File: 2073. Time Needed to Buy Tickets/TimeRequiredToBuy.py
from typing import List


class Solution:
    
    # O(n^2) time,
    # O(1) space,
    # Approach: simulation, 
    def timeRequiredToBuy(self, tickets: List[int], k: int) -> int:
        target_ticket = tickets[k]
        time = 0
        
        for i in range(target_ticket-1):
            for j in range(len(tickets)):
                if tickets[j] > 0:
                    time += 1
                    tickets[j] -= 1
        
        for i in range(k+1):
            if tickets[i] > 0:
                time += 1
                tickets[i] -= 1
                
        return time
    

    # O(n) time,
    # O(1) space,
    # Approach: math, 
    def timeRequiredToBuy2(self, tickets: List[int], k: int) -> int:
        time = 0
        
        for i, curr_ticket in enumerate(tickets):
            time += min(tickets[k], curr_ticket) if i <= k else min(tickets[k]-1, curr_ticket) 
                
        return time

File: 2073. Time Needed to Buy Tickets/time_required_to_buy.rs
use std::cmp::min;

impl Solution {
    // O(n) time,
    // O(1) space,
    // Approach: array,
    pub fn time_required_to_buy(tickets: Vec<i32>, k: i32) -> i32 {
        let mut time = 0;
        let rounds = tickets[k as usize];
        let n = tickets.len();

        for i in 0..n {
            if i <= k as usize {
                let rounds_before_leaving = min(rounds, tickets[i]);
                time += rounds_before_leaving;
            } else {
                let rounds_before_leaving = min(rounds - 1, tickets[i]);
                time += rounds_before_leaving;
            }
        }

        time
    }
}
